### PR TITLE
Test Run Title and Publish Test Attachments

### PR DESCRIPTION
**Pull request by prawalagarwal**
Description: The PublishTestResults Task's xplat implementation is currently different from windows implementation. Users choice of Run Title and option to not publish result files do not exist in the current xplat implementation.

Solution: Changes have been made for Junit, xunit and nunit to honor users choice of run ttitle and user can now opt out of publishing test attachments. Link to vso-agent-tasks: https://github.com/Microsoft/vso-agent-tasks/pull/954. Link to vso-task-lib changes:  https://github.com/Microsoft/vso-task-lib/pull/19

Testing: Manual testing for now. Appropriate unit tests will be added if required.

### CODE REVIEW

File: src/agent/testresultreader.ts [hunk 1]
```diff
@@ -119,39 +120,64 @@ export class ResultReader {
...
                     }
                 }
-                if(numTestSuites > 1) {
-                    runSummary.name = "JUnit";
-                }
+				
+				if (!(runTitle && 0 !== runTitle.length)) {
+					if(numTestSuites > 1) {
+						if (fileName && 0 !== fileName.length) {
+							runSummary.name = "JUnit" + fileName;
```
Comment (kaadhina) [inline src/agent/testresultreader.ts]: should we add a space between 'junit' and the file name?

File: src/agent/commands/results.publish.ts [hunk 2]
```diff
@@ -45,7 +48,10 @@ export class ResultsPublishCommand implements cm.IAsyncCommand {
             releaseEnvironmentUri: this.executionContext.variables["release.environmentUri"],
             releaseUri: this.executionContext.variables["release.releaseUri"],
             platform: platform,
-            config: config
+            config: config,
+			runTitle: runTitle,
+			fileNumber: fileNumber,
```
Comment (kaadhina) [inline src/agent/commands/results.publish.ts]: Can we avoid the filenumber parameter and send it along with runtitle? we can avoid duplicating the logic in the readers.

File: src/agent/testresultreader.ts [hunk 3]
```diff
@@ -119,39 +120,64 @@ export class ResultReader {
...
+						if (fileName && 0 !== fileName.length) {
+							runSummary.name = "JUnit" + fileName;							
+						}
+						else {
+							if (fileNumber == 0)
+							{
+								runSummary.name = "JUnit";
+							}
+							else {
+								runSummary.name = "JUnit" + fileNumber;
```
Comment (kaadhina) [inline src/agent/testresultreader.ts]: space here also

File: src/agent/testrunpublisher.ts [hunk 4]
```diff
@@ -4,6 +4,7 @@ import ctxm = require('./context');
 import cm = require('./common');
```
Comment (kaadhina) [inline src/agent/testrunpublisher.ts]: The tests we have today are very basic. If we re going to make the js implementation as default one, we should definitely add more tests. Even if not as a part of this story, may be we take this up as a separate work item.

File: src/agent/testresultreader.ts (hunk 1 above)
Comment (prawalagarwal) [inline src/agent/testresultreader.ts]: Yes

File: src/agent/commands/results.publish.ts (hunk 2 above)
Comment (prawalagarwal) [inline src/agent/commands/results.publish.ts]: We need the file number for attaching to default titles too. Then we would have to extract it out of run title. I think this is fine given our requirements.

File: src/agent/testresultreader.ts (hunk 3 above)
Comment (prawalagarwal) [inline src/agent/testresultreader.ts]: Yes

File: src/agent/commands/results.publish.ts [hunk 5]
```diff
@@ -37,6 +37,9 @@ export class ResultsPublishCommand implements cm.IAsyncCommand {
 
         var platform: string = this.command.properties['platform'];
         var config : string = this.command.properties['config'];
+		var runTitle : string = this.command.properties['runTitle'];
```
Comment (allendm-msft) [inline src/agent/commands/results.publish.ts]: fix the formatting

File: src/agent/commands/results.publish.ts [hunk 6]
```diff
@@ -59,12 +65,11 @@ export class ResultsPublishCommand implements cm.IAsyncCommand {
             reader = new trr.XUnitResultReader();
         }
         else {
-            this.command.warning("Test results of format '" + resultType + "'' are not supported by the VSO/TFS OSX and Linux build agent");
+            this.command.warning("Test results of format '" + resultType + "'' are not supported by the VSTS/TFS OSX and Linux build agent");
```
Comment (allendm-msft) [inline src/agent/commands/results.publish.ts]: remove references to OSX/Linux.. this is supposed to run anywhere

File: src/agent/testresultreader.ts [hunk 7]
```diff
@@ -119,39 +120,64 @@ export class ResultReader {
...
-                    runSummary.name = "JUnit";
-                }
+				
+				if (!(runTitle && 0 !== runTitle.length)) {
+					if(numTestSuites > 1) {
+						if (fileName && 0 !== fileName.length) {
+							runSummary.name = "JUnit" + fileName;							
+						}
+						else {
+							if (fileNumber == 0)
```
Comment (allendm-msft) [inline src/agent/testresultreader.ts]: runSummary.Name = "JUnit";
if(fileNumber > 0) runSummary.Name = runSummary.Name + " " + fileNumber

File: src/agent/testresultreader.ts [hunk 8]
```diff
@@ -183,19 +209,40 @@ export class ResultReader {
...
+		}
+        else if(rootNode.attributes().name) {
             testSuiteSummary.name = rootNode.attributes().name;
         }
+		else {
+			if (fileName && 0 !== fileName.length) {				
+				testSuiteSummary.name = "JUnit" + fileName;							
+			}
+			else {
+				if (fileNumber == 0)
```
Comment (allendm-msft) [inline src/agent/testresultreader.ts]: duplicate code.. extract into common function

File: src/agent/commands/results.publish.ts (hunk 6 above)
Comment (prawalagarwal) [inline src/agent/commands/results.publish.ts]: This is for VSTest type. Hence this warning is required.

File: src/agent/testresultreader.ts [hunk 9]
```diff
@@ -456,23 +521,42 @@ export class ResultReader {
...
+			runTitle = runContext.runTitle;
+		    fileNumber = runContext.fileNumber;
 
             //Release uri, Release environment uri
             var releaseUri = runContext.releaseUri;
             var releaseEnvironmentUri = runContext.releaseEnvironmentUri;
         }
 
-        var runName = "XUnit Test Run";
+		if (runTitle && 0 !== runTitle.length) {
```
Comment (allendm-msft) [inline src/agent/testresultreader.ts]: this code is similar across all readers.. refactor

File: src/agent/testrunpublisher.ts [hunk 10]
```diff
@@ -68,24 +69,7 @@ export class TestRunPublisher {
...
-            utilities.readFileContents(resultFilePath, "ascii").then(function (res) {
-                var contents = new Buffer(res).toString('base64');
-                _this.service.createTestRunAttachment(createdTestRun.id, path.basename(resultFilePath), contents).then(
-                    function (attachment) {
-                        defer.resolve(createdTestRun);
-                    },
-                    function (err) {
-                        // We can skip attachment publishing if it fails to upload
-                        if (_this.command) {
-                            _this.command.warning("Skipping attachment : " + resultFilePath + ". " + err.statusCode + " - " + err.message);
```
Comment (allendm-msft) [inline src/agent/testrunpublisher.ts]: this param is no longer used here. remove it.

**Comment by allendm-msft** [comment]
1. fix the formatting across all the files. lets keep the code clean
2. refactor the new code for reuse across the filereaders
3. add more tests

File: src/agent/testresultreader.ts (hunk 8 above)
Comment (prawalagarwal) [inline src/agent/testresultreader.ts]: Done

File: src/agent/testresultreader.ts (hunk 9 above)
Comment (prawalagarwal) [inline src/agent/testresultreader.ts]: Done

File: src/agent/testrunpublisher.ts (hunk 4 above)
Comment (prawalagarwal) [inline src/agent/testrunpublisher.ts]: I have added a few tests. Will add more eventually.

File: src/test/publishertests.ts [hunk 11]
```diff
@@ -150,31 +167,356 @@ describe('PublisherTests', function() {
...
+            config: "",
+            runTitle: "My Title",
+            publishRunAttachments: true,
+            fileNumber: "9",
+            releaseUri: "abc",
+            releaseEnvironmentUri: "xyz"
+        };
+        var resultsFile = path.resolve(__dirname, './testresults/xunitresults.xml');
+        var reader = new trr.XUnitResultReader();
+        //var testRunWithResults = reader.readResults(resultsFile, runContext);
```
Comment (allendm-msft) [inline src/test/publishertests.ts]: remove commented out code

File: src/test/publishertests.ts [hunk 12]
```diff
@@ -150,31 +167,356 @@ describe('PublisherTests', function() {
...
+        function(err) {
+            assert(false, 'XUnit Reader Failed! Details : ' + err.message);
+            done();
+        });		
+    })
+    
+    it('results.publish : XUnit reader sanity check with run title and file number', function(done) {
+        
+        var testRun;
+        var runContext: trp.TestRunContext = {
```
Comment (allendm-msft) [inline src/test/publishertests.ts]: dont you feel refactoring into a createtestruncontext with few params is a better option than copy/pasting it ?

File: src/test/publishertests.ts [hunk 13]
```diff
@@ -150,31 +167,356 @@ describe('PublisherTests', function() {
...
+            //Verifying the test run details
+            assert.strictEqual("My Title 9", testRun.name);			
+            done();
+        },
+        function(err) {
+            assert(false, 'XUnit Reader Failed! Details : ' + err.message);
+            done();
+        });		
+    })
 });
```
Comment (allendm-msft) [inline src/test/publishertests.ts]: in which of our builds will you add these tests?

**Comment by allendm-msft** [comment]
Approved with suggestions

**Pull request merged**